File: src/main/java/introkotlin/samplejava/step3/Money.java
package introkotlin.samplejava.step3;

import java.util.Objects;

/**
 * Le bon Value Object en Java.
 */
public final class Money {

    private final double value;

    private final String currency;

    public Money(double value, String currency) {
        this.value = value;
        this.currency = currency;
    }

    public double getValue() {
        return value;
    }

    public String getCurrency() {
        return currency;
    }

    @Override
    public String toString() {
        return "Money(value=" + value + ", currency=" + currency + ")";
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        final var other = (Money) obj;
        return other.value == value && Objects.equals(currency, other.currency);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value, currency);
    }

}

File: src/main/kotlin/introkotlin/samplekotlin/step3/Money.kt
package introkotlin.samplekotlin.step3

import java.util.*

/**
 * On a ici gratuitement... un constructeur,
 * deux propriétés et leurs getters associés.
 *
 * On supprime le caractère ouvert de la classe : on
 * ne peut plus en hériter.
 */
class Money(
  val value: Double,
  val currency: String
) {

  /**
   * Je peux invoquer du Java de manière transparente.
   */
  override fun hashCode(): Int {
    return Objects.hash(value, currency)
  }

  override fun equals(other: Any?): Boolean {
    // Le when pour faire une sorte de switch
    // On a également le compilateur qui détecte automatiquement
    // le type de other pour la branche "other is Money". Pas
    // besoin de caster
    return when {
      this === other -> true
      other is Money -> value == other.value && currency == other.currency
      else -> false
    }
  }

  /**
   * Autre syntaxe pour implémenter une méthode, + template string.
   */
  override fun toString() = "Money(value=$value, currency=$currency)"

}
